packages/scheduler/lib/workers/scheduling/scheduling.ts

```diff
@@ -0,0 +1,48 @@
+import type knex from 'knex';
+import type { Schedule } from '../../types.js';
+import { DbSchedule, SCHEDULES_TABLE } from '../../models/schedules.js';
+import type { Result } from '@nangohq/utils';
+import { Err, Ok, stringifyError } from '@nangohq/utils';
+import { TASKS_TABLE } from '../../models/tasks.js';
+
+export async function dueSchedules(db: knex.Knex): Promise<Result<Schedule[]>> {
+    try {
+        const query = db
+            .with(
+                'last_run_times',
+                // calculate the last run time for each schedule that is started/not deleted
+                db
+                    .select(
+                        's.id',
+                        db.raw(`
+                            s.starts_at + (FLOOR(EXTRACT(EPOCH FROM (NOW() - s.starts_at)) / EXTRACT(EPOCH FROM s.frequency)) * s.frequency) AS last_run_time
```

Comment:
Would it be easier to just store this when you run a task? at least it would reflect actual usage not idealistic past

Reply:
1. it is supposed to be the expected due time, not the actual one. Let me rename the CTE and field to make it clearer
2. There is a tradeoff between runtime vs pre-computing: runtime is more expensive but with pre-computing you need to ensure the pre-computation always happen successfully and potential mutation error must be handled. I would prefer to stick with runtime calculation until we hit scaling issue.